### PR TITLE
Conjugate gradient

### PR DESCRIPTION
**Pull request by joanrue**
Added the Conjugate Gradient Method iterative solver based on the Sepand's Solver.py base class.

### CODE REVIEW

**Comment by joanrue** [comment]
My last commit should have been: 
"[solver/CG] Refactor to leverage base classes' functionalities"
"Stopping criterion is an instance of opt.stop.AbsError instead of created from scratch. Fit function is called from base class instead of defining it from scratch. Also added a broadcastable test for b and x0"

File: src/pycsou/abc/solver.py [hunk 1]
```diff
@@ -566,6 +566,7 @@ def _m_persist(self):
 
     class _Worker(threading.Thread):
         def __init__(self, solver: "Solver"):
+            super().__init__()
```
Comment (SepandKashani) [inline src/pycsou/abc/solver.py:569]: ok based on our discussion.

File: src/pycsou/opt/solver/cg.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+class CG(pycs.Solver):
+    r"""
+    Conjugate Gradient Method.
+    The Conjugate Gradient method solves the system of linear equations of the form
+    .. math::
+       {\mathbf{A}\mathbf{x} = \mathbf{b}},
+    for the vector :math:`\mathcal{x}: \mathbb{x}^N`.
+    where:
+    * :math:`\mathcal{A}: \mathbb{R}^M\times^N` is a *symmetric* *positive definite* matrix.
```
Comment (SepandKashani) [inline src/pycsou/opt/solver/cg.py]: *positive def* OPERATOR, not necessarily a matrix. (We only need to know how to apply the op afterall.)

File: src/pycsou/opt/solver/cg.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+            log_var=log_var,
+        )
+
+        self._a = a
+        self._b = pycrt.coerce(b)
+
+    def fit(
+        self,
+        x0: pyct.NDArray = None,
+        stop_crit: pycs.StoppingCriterion = None,
```
Comment (SepandKashani) [inline src/pycsou/opt/solver/cg.py]: If `tol` is provided in `fit()`, then you should not have `stop_crit` in the signature anymore.
It is the job of `fit()` to construct the AbsErr stop_crit and call `super().fit()` with it.

File: src/pycsou/opt/solver/cg.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        mode: pycs.Mode = pycs.Mode.BLOCK,
+        tol: float = 1e-5,
+    ):
+        r"""
+        Solve the minimization problem defined in :py:meth:`CG.__init__`, with the provided
+        run-specific parameters.
+        Parameters
```
Comment (SepandKashani) [inline src/pycsou/opt/solver/cg.py:55]: Add an empty line before `Parameters`.

File: src/pycsou/opt/solver/cg.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        ----------
+        x0: NDArray
+           (..., N) Starting guess(es) for the solution(s). Defaults to a zero NDArray if unspecified.
```
Comment (SepandKashani) [inline src/pycsou/opt/solver/cg.py]: Prefer the term "initial point" to "starting guess".

File: src/pycsou/opt/solver/cg.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+        self._mstate["p"] = r.copy()
+
+    def m_step(self):
+        mst = self._mstate
+        x = mst["x"]
+        r = mst["r"]
+        p = mst["p"]
+
+        ap = self._a.apply(p)
+        rr = (r * r).sum(-1)[..., None]
```
Comment (SepandKashani) [inline src/pycsou/opt/solver/cg.py]: `rr = xp.linalg.norm(r, ord=2, axis=-1, keepdims=True` should work and avoids 2 temporary arrays in the process.

File: src/pycsou/opt/solver/cg.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+        alpha = rr / (p * ap).sum(-1)[..., None]
```
Comment (SepandKashani) [inline src/pycsou/opt/solver/cg.py]: `rr / (p*ap).sum(axis=-1, keepdims=True)` avoids having to do the `[...,None]` indexing at the end.

File: src/pycsou/opt/solver/cg.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+        x = x + alpha * p
+        r = r - alpha * ap
+        beta = (r * r).sum(-1)[..., None] / rr
```
Comment (SepandKashani) [inline src/pycsou/opt/solver/cg.py]: Same comment as above.

**Review by SepandKashani — CHANGES_REQUESTED** [review]

File: src/pycsou/opt/solver/cg.py [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+
+class CG(pycs.Solver):
+    r"""
+    Conjugate Gradient Method.
+    The Conjugate Gradient method solves the system of linear equations of the form
+    .. math::
+       {\mathbf{A}\mathbf{x} = \mathbf{b}},
+    for the vector :math:`\mathcal{x}: \mathbb{x}^N`.
+    where:
+    * :math:`\mathcal{A}: \mathbb{R}^M\times^N` is a *symmetric* *positive definite* operator.
```
Comment (SepandKashani) [inline src/pycsou/opt/solver/cg.py]: You use both M and N as indices here, but A must be square.
Double-check the whole docstring and standardize on one letter only.

Also, rewrite this as `\mathcal{A}: \mathbb{R}^{N} \to \mathbb{R}^{N}`

File: src/pycsou/opt/solver/cg.py [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+    * :math:`\mathcal{b}: \mathbb{b}^M`.
+    """
+
+    def __init__(
+        self,
+        a: pyco.PosDefOp,
+        b: pyct.NDArray,
+        log_var: pyct.VarName = ("x",),
```
Comment (SepandKashani) [inline src/pycsou/opt/solver/cg.py:32]: I missed this last time: why are `folder`, `exist_ok`, `writeback_rate`, `verbosity` missing here?

File: src/pycsou/opt/solver/cg.py (hunk 9 above)
Comment (SepandKashani) [reply src/pycsou/opt/solver/cg.py]: In `CG.fit()`, the docstring says "the minimization problem defined in :py:meth:`CG.__init__`", but Ax=b is not a minimization problem.
You should instead write \argmin_{x\in\bR^{N}} ||Ax - b||_{2}^{2}, where A is ...

File: src/pycsou/opt/solver/cg.py [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+                    f"(array module {pycu.get_array_module(x0)} is different from {xp}."
+                )
+        else:
+            x0 = xp.zeros((1, self._b.shape[-1]), dtype=pycrt.getPrecision().value)
+
+        return super().fit(x0, mode=mode, stop_crit=stop_crit)
+
+    def m_init(self, x0: pyct.NDArray, tol: float = 1e-5):
+        self._mstate["x"] = x0
+        self._mstate["r"] = r = self._b - self._a.apply(x0)
```
Comment (SepandKashani) [inline src/pycsou/opt/solver/cg.py]: `= r =` is useless.

File: src/pycsou/opt/solver/cg.py [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+        ap = self._a.apply(p)
+        rr = xp.linalg.norm(r, ord=2, axis=-1, keepdims=True) ** 2
+        alpha = rr / (p * ap).sum(axis=-1, keepdims=True)
+        x = x + alpha * p
+        r = r - alpha * ap
+        beta = xp.linalg.norm(r, ord=2, axis=-1, keepdims=True) ** 2 / rr
+        p = r + beta * p
+        mst["x"], mst["r"], mst["p"] = x, r, p
+
+    @pycrt.enforce_precision(o=True)
```
Comment (SepandKashani) [inline src/pycsou/opt/solver/cg.py]: No decorator should be used here. If your `m_init` and `m_step` are correct, the output dtype should match `b` and `x0`s dtype.

File: src/pycsou/opt/solver/cg.py [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+           * MANUAL: fit() + steps()
+        tol: Real
+           Tolerance for convergence, norm(residual) <= tol.  Defaults to :math:`1e-5` if unspecified.
+        """
+        # Create a stopping criteria from opt, using tol, and then call super().fit() method
+
+        xp = pycu.get_array_module(self._b)
+        x0 = x0 if (x0 is None) else pycrt.coerce(x0)
+
+        try:
```
Comment (SepandKashani) [inline src/pycsou/opt/solver/cg.py]: `AbsError` will check `tol > 0`: don't need to test again here.

File: src/pycsou/opt/solver/cg.py [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+                )
+            try:
+                assert pycu.get_array_module(x0) == xp
+            except:
+                raise ValueError(
+                    f"Input initial guess has a mismatch in its shape array module with data array `b` "
+                    f"(array module {pycu.get_array_module(x0)} is different from {xp}."
+                )
+        else:
+            x0 = xp.zeros((1, self._b.shape[-1]), dtype=pycrt.getPrecision().value)
```
Comment (SepandKashani) [inline src/pycsou/opt/solver/cg.py]: `dtype = self._b.dtype`.

File: src/pycsou/opt/solver/cg.py [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+
+import pycsou.abc.operator as pyco
+import pycsou.abc.solver as pycs
+import pycsou.opt.stop as pycos
+import pycsou.runtime as pycrt
+import pycsou.util as pycu
+import pycsou.util.ptype as pyct
+
+
+def is_broadcastable(arr1, arr2):
```
Comment (SepandKashani) [inline src/pycsou/opt/solver/cg.py]: Use `numpy.broadcast_shapes()` instead of rolling out your own function.

File: src/pycsou/opt/solver/cg.py [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+                assert pycu.get_array_module(x0) == xp
+            except:
+                raise ValueError(
+                    f"Input initial guess has a mismatch in its shape array module with data array `b` "
+                    f"(array module {pycu.get_array_module(x0)} is different from {xp}."
+                )
+        else:
+            x0 = xp.zeros((1, self._b.shape[-1]), dtype=pycrt.getPrecision().value)
+
+        return super().fit(x0, mode=mode, stop_crit=stop_crit)
```
Comment (SepandKashani) [inline src/pycsou/opt/solver/cg.py:94]: I now have a doubt on this. Leave as-is and see if it runs.

**Review by SepandKashani — CHANGES_REQUESTED** [review]
Looking better. Still some things to tackle.

**Review by SepandKashani — APPROVED** [review]

**Pull request merged**